flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/SorterFactory.java

```diff
@@ -0,0 +1,174 @@
+	 * @param memory
+	 * @return
+	 */
+	public <T> InMemorySorter<T> createSorter(ExecutionConfig config, TypeSerializer<T> serializer, TypeComparator<T> comparator, List<MemorySegment> memory) {
+		if (config.isCodeGenerationForSorterEnabled()){
+			try {
+				return createCodegenSorter(serializer, comparator, memory);
+			} catch (IOException | TemplateException | ClassNotFoundException | IllegalAccessException |
+					InstantiationException | NoSuchMethodException | InvocationTargetException | CompileException e) {
+				String msg = "Serializer: " + serializer +
+						"[" + serializer + "], comparator: [" + comparator + "], exception: " + e.toString();
+				if (!forceCodeGeneration) {
+					LOG.warn("An error occurred while trying to create a code generated sorter. Using non-codegen sorter instead. " + msg);
+					return createNonCodegenSorter(serializer, comparator, memory);
+				} else {
+					throw new RuntimeException("An error occurred while trying to create a code generated sorter. Failing the job, because forceCodeGeneration. " + msg);
+				}
+			}
+		} else {
+			return createNonCodegenSorter(serializer, comparator, memory);
+		}
+	}
+
+	private <T> InMemorySorter<T> createCodegenSorter(TypeSerializer<T> serializer, TypeComparator<T> comparator, List<MemorySegment> memory)
+			throws IOException, TemplateException, ClassNotFoundException, IllegalAccessException, InstantiationException, NoSuchMethodException, InvocationTargetException, CompileException {
+		SorterTemplateModel sorterModel = new SorterTemplateModel(comparator);
+
+		Constructor sorterConstructor;
+
+		synchronized (this){
+			if (constructorCache.getOrDefault(sorterModel.getSorterName(), null) != null) {
+				sorterConstructor = constructorCache.get(sorterModel.getSorterName());
+			} else {
+				String sorterName = sorterModel.getSorterName();
+				String generatedCode = this.templateManager.getGeneratedCode(sorterModel);
+				this.classCompiler.cook(generatedCode);
+
+				sorterConstructor = this.classCompiler.getClassLoader().loadClass(sorterName).getConstructor(
+						TypeSerializer.class, TypeComparator.class, List.class
+				);
+
+				constructorCache.put(sorterName, sorterConstructor);
+			}
+		}
+
+		@SuppressWarnings("unchecked")
+		InMemorySorter<T> sorter = (InMemorySorter<T>) sorterConstructor.newInstance(serializer, comparator, memory);
+
+		if (LOG.isInfoEnabled()){
+			LOG.info("Using a custom sorter : " + sorter.toString());
+		}
+
+		return sorter;
+	}
+
+	private <T> InMemorySorter<T> createNonCodegenSorter(TypeSerializer<T> serializer, TypeComparator<T> comparator, List<MemorySegment> memory) {
+		InMemorySorter<T> sorter;
+		// instantiate a fix-length in-place sorter, if possible, otherwise the out-of-place sorter
+		if (comparator.supportsSerializationWithKeyNormalization() &&
+				serializer.getLength() > 0 && serializer.getLength() <= THRESHOLD_FOR_IN_PLACE_SORTING &&
+				comparator.isNormalizedKeyPrefixOnly(comparator.getNormalizeKeyLen())) {
```

Comment:
> So the additional condition is fixing a bug?

Yes, but currently the `FixedLengthRecordSorter` is never used, so that is why this hasn't manifested yet (see https://issues.apache.org/jira/browse/FLINK-4705)

> Shouldn't the condition should be inverted

Yes, sorry, I've fixed the condition in b2bf565edda6ada763c27c132cde58b0efd1fe7b .